### ISSUE TITLE
prefer-spread flagging `Array#concat(…)` for non-iterable objects

### ISSUE THREAD

**Issue by bdunks** [sindresorhus/eslint-plugin-unicorn#1064]
After upgrading to `27.0.0`, I'm finding that the `prefer-spread` rule is raising `ERROR_ARRAY_CONCAT` errors when a `concat` method exists on non-iterable objects.

Specifically, this I've witnessed this occurring when chaining an ApolloLink for the ApolloClient.  The ApolloLink object defines a static `concat` method.

```js
const client = new ApolloClient({
  link: authLink.concat(httpLink),  //<-- Raises `Array#concat(…)` error
  cache: new InMemoryCache({}),
});
```

In this case, authLink is of type ApolloLink, which has the following type definition:
```js
export declare class ApolloLink {
    static empty(): ApolloLink;
    static from(links: (ApolloLink | RequestHandler)[]): ApolloLink;
    static split(test: (op: Operation) => boolean, left: ApolloLink | RequestHandler, right?: ApolloLink | RequestHandler): ApolloLink;
    static execute(link: ApolloLink, operation: GraphQLRequest): Observable<FetchResult>;
    static concat(first: ApolloLink | RequestHandler, second: ApolloLink | RequestHandler): ApolloLink;
    constructor(request?: RequestHandler);
    split(test: (op: Operation) => boolean, left: ApolloLink | RequestHandler, right?: ApolloLink | RequestHandler): ApolloLink;
    concat(next: ApolloLink | RequestHandler): ApolloLink;
    request(operation: Operation, forward?: NextLink): Observable<FetchResult> | null;
    protected onError(error: any, observer?: ZenObservable.Observer<FetchResult>): false | void;
    setOnError(fn: ApolloLink["onError"]): this;
}
```

**Comment by fisker** [comment]
Sorry, we can't know the type of `authLink`

**Comment by sindresorhus** [comment]
The only way we can improve this is with https://github.com/sindresorhus/eslint-plugin-unicorn/issues/347.

**Comment by bdunks** [comment]
Hi @fisker  - I'm sorry, I should have posted an example that doesn't rely on a third party library.  Hopefully it's okay that this is illustrated as a react component; I believe it gets the point across.


```js
class Chainable {
  concat(newLink: Chainable){
    console.log("I don't do what you expect");
  }
}

export const Component = () => {
  const notAnArray = new Chainable();
  const alsoNotAnArray = new Chainable();
  const chainedItem = notAnArray.concat(alsoNotAnArray); //<-- Raises `Array#concat(…)` error
  return (
    <></>
  );
}
```

I've found that adding `object: 'Array',` to [the rule definition's `method selector`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/d707e839d4e85193a8be8b9b2d9a73137773915a/rules/prefer-spread.js#L32) will prevent the issue (similar to the `arrayFromCallSelector` just above), but I don't know the broader implications of this change.

**Comment by fisker** [comment]
Adding `object: 'Array'` means the selector only matches `Array.concat()` it's for static methods.

**Comment by bdunks** [comment]
Got it, thank you.  

Do you agree it doesn't make sense to error on arbitrary objects that cannot be spread?

If we can't validate that the object is of type array, could we consider adding explicit ignore options similar the `catch-error-name` rule, checking the `node.callee.object.name` against the ignore list?

This might allow a rule definition like this, which would be helpful when we want the rule enabled in general, but have isolated cases that a third party library defines classes with a concat method:

```js
            'unicorn/prefer-spread': [
                'error',
                {
                    ignore: [/^authLink/i],
                },
            ],
```

**Comment by JounQin** [comment]
Or can we add an option prefer `concat`?

```
// better to me, specially for ts files
const singleOrArr = xx

const arr2 = arr1.concat(item, arr2, singleOrArr)

const arr3 = [...arr1, item, ...arr2, ...(Array.isArray(singleOrArr) ? singleOrArr : [singleOrArr])]
```

**Comment by fisker** [comment]
We may consider and option to ignore `.concat()`, but absolutely not suggesting the opposite.

**Comment by fisker** [comment]
FYI: I just refactor Prettier codebase [from `.concat()` to `...` ](https://github.com/prettier/prettier/pull/10129), it's much more clear that the argument is an array or not, and during that refactor, I realized `Array.isArray` test is useful suggestion, and already added to the rule #1054

**Comment by JounQin** [comment]
> We may consider and option to ignore `.concat()`, but absolutely not suggesting the opposite.

Yeah, I mean an option to ignore `concat`, I think it's useful personally.

**Comment by bdunks** [comment]
I've spent a bit more time digging in, and I'm wondering [if this conditional/check should be changed](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/da685f7409669728bacf73e7edb1d283eb02c14f/rules/prefer-spread.js#L261) to:

```js
if (!staticResult || !Array.isArray(staticResult.value)) {
	return;
}
```

The rule currently bails out if it a `staticResult` is not an Array.  However, it doesn't bail if there is no `staticResult`.  In the case of an arbitrary object, `staticResult` is null.  

This would also address #1068 without relying on the `callee.object.name` starting with a capital letter.

Note that this causes a number of existing tests to fail (e.g., `'foo.concat(2)'`, is no longer an invalid case), however, perhaps those test should also more explicitly ensure we are testing arrays.  The following case is still invalid:

```js
outdent`
	const foo= [];
	foo.concat(2);
`,
```

**Comment by fisker** [comment]
>  I'm wondering if this conditional/check should be changed to:

No, we can't,  most time `staticResult` can't be  calculated, even this simple code

```js
class A {};

const foo = [1, 2, new A()];
```

The `staticResult` is `null`

**Comment by moltar** [comment]
This happens with `Buffer` too, which is a built in.

**Comment by Slapbox** [comment]
Just hit this myself. Seems like a shame to need to disable `prefer-spread` entirely to avoid this. It would be great it the rule excluded `concat`.

**Comment by the-ult** [comment]
hmm.
Could this be an issue with a method that returns a `string` as well?

![Screenshot 2021-12-09 at 20 23 27](https://user-images.githubusercontent.com/4863062/145462223-c8a195aa-3ad2-49eb-bc10-54a7dc76efb4.png)

while this isn't a problem:
```ts
test(){
   const myTestString = 'hello';
    const result = myTestString.concat("this is a test");
  }

```

**Comment by the-ult** [comment]
The above issue with `string.concat` still seems to appear. Is this known/correct?

```ts

            let test = 'foo;
            test = test.concat('bar');
            //          ^ Prefer the spread operator over `Array#concat(…)`.eslint[unicorn/prefer-spread]
```

**Comment by FunctionDJ** [comment]
Until #347, I worked around it by just renaming my custom `concat` method to `concatena`.
